File: back_end/modules.py
"""
Scenario configuration — the Flight Deck 2050 modules as metric definitions.

Impacts are now RATINGS (Very Low / Low / Medium / High / Very High) on three
dimensions, built with ``rate(...)``:

    rate(climate="High", financial="Medium", time="Low")

Rules of thumb while filling this in:
  * Higher is always better, incl. time (Very High time = fastest to deliver).
  * Omit a dimension to mark it "not applicable" (skipped in scoring) — see the
    funding module, which is climate-neutral.
  * Sliders (Percentage / Level) interpolate between their two rated ends, so you
    only rate the low end and the high end.

Every rating below is a PLACEHOLDER marked # TODO — replace with your judgement.
"""

from __future__ import annotations

from .impacts import rate
from .metrics import AllocationMetric, ChoiceMetric, LevelMetric, PercentageMetric
from .scoring import ScoreWeights

# Handy placeholders. Replace per-dimension as you go.
_BASE = rate("Very Low", "Very Low", "Very Low")            # a 'do nothing' low end
_TODO = rate("Medium", "Medium", "Medium")                  # TODO: set real high-end ratings

MODULES: list = [

    # MOD 01 - CO2 cuts - SAF
    PercentageMetric(
        id="saf_co2_cuts",
        label="SAF CO2 Cuts (%)",
        impact_at_0=rate(climate="Low", financial="Very High", time="High"),
        impact_at_100=rate(climate="High", financial="Very Low", time="High"),
        default_pct=50.0,
    ),

    # MOD 02 - Time to 100% SAF - SAF
    PercentageMetric(
        id="saf_time_to_100",
        label="Time to 100% SAF (% vol)",
        impact_at_0=rate(climate="Very Low", financial="Medium", time="Very Low"),
        impact_at_100=rate(climate="Very High", financial="Medium", time="Very High"),
        default_pct=50.0,
    ),

    # MOD 03 - Willingness to contribute financially - SAF
    PercentageMetric(
        id="saf_WTC",
        label="Willingness to contribute financially (% premium accepted)",
        impact_at_0=rate(climate="Low", financial="Low"),
        impact_at_100=rate(climate="High", financial="High"),
        default_pct=50.0,
    ),

    # MOD 04 - System maintenance effort - Bio-Sensors
    # Level 1 = Low effort ... Level 5 = High effort. The ratings below define
    # what each end MEANS for climate / financial / time.
    LevelMetric(
        id="biosensor_maintenance",
        label="System Maintenance Effort",
        impact_at_min=rate(climate="Very Low", financial="Low"),
        impact_at_max=rate(climate="Very High", financial="High"),
    ),

    # MOD 05 - Reliability gain vs. current sensors - Bio-Sensors
    PercentageMetric(
        id="biosensor_reliability",
        label="Reliability Gain vs. Current Sensors (% more reliable)",
        impact_at_0=rate(climate="Very Low", financial="Very Low", time="High"),
        impact_at_100=rate(climate="Very High", financial="Very High", time="Low"),
        default_pct=50.0,
    ),

    # MOD 06 - Certification support - Bio-Sensors
    LevelMetric(
        id="biosensor_certification",
        label="Certification Support",
        impact_at_min=rate(climate="Very Low", time="Very Low"),
        impact_at_max=rate(climate="Very High", time="Very High"),
    ),

    # MOD 07 - Data Depth - Digital Passport
    PercentageMetric(
        id="passport_digital_depth",
        label="Digital Depth (%)",
        impact_at_0=rate(climate="Low", financial="High", time="Low"),
        impact_at_100=rate(climate="High", financial="Medium", time="High"),
        default_pct=50.0,
    ),

    # MOD 08 - Regulatory Support - Digital Passport
    LevelMetric(
        id="passport_reg_support",
        label="Regulatory Support",
        impact_at_min=rate(climate="Medium", financial="Very Low", time="Very Low"),
        impact_at_max=rate(climate="High", financial="Very High", time="Very High"),
    ),

    # MOD 09 - Industry Adoption - Materials Digital Passport
    LevelMetric(
        id="passport_adoption",
        label="Industry Adoption",
        impact_at_min=rate(climate="Low", financial="Low", time="Very Low"),
        impact_at_max=rate(climate="High", financial="High", time="Very High"),
    ),

    # MOD 10 - Funding payback period — REMOVED: replaced by MOD 19 (funding_split)
    # in the Emission Guide & Eco-Label group, per design decision.

    # MOD 11 - Global Participation (incl. eco-labels) - Emission Guide & Eco-label
    LevelMetric(
        id="EGE_participation",
        label="Global Participation (incl. Eco-labels)",
        impact_at_min=rate(climate="Very Low", financial="Very Low", time="Very High"),
        impact_at_max=rate(climate="Very High", financial="Very High", time="Very Low"),
    ),

    # MOD 12 - Market trust in label - Emission Guide & Eco-label
    LevelMetric(
        id="EGE_market_trust",
        label="Market Trust in Label",
        impact_at_min=rate(climate="Low", time="Very Low"),
        impact_at_max=rate(climate="High", time="Very High"),
    ),

    # MOD 13 - Sustainability mode shift - Transport of Aircraft Elements
    PercentageMetric(
        id="transport_mode_shift",
        label="Shift to Low Climate, Resource, and Air-Quality Impact Modes (%)",
        impact_at_0=rate(climate="Very Low", financial="Very Low", time="Very Low"),
        impact_at_100=rate(climate="Very High", financial="Very High", time="Very High"),
        default_pct=50.0,
    ),

    # MOD 14 - Carbon tax - Transport of Aircraft Elements
    PercentageMetric(
        id="transport_carbon_tax",
        label="Carbon Tax (%)",
        impact_at_0=rate(climate="Low", financial="Low", time="Very High"),
        impact_at_100=rate(climate="High", financial="High", time="Very Low"),
        default_pct=50.0,
    ),

    # MOD 15 - Schedule buffer - Transport of Aircraft Elements
    LevelMetric(
        id="transport_buffer",
        label="Schedule Buffer",
        impact_at_min=rate(climate="Very Low", time="High"),
        impact_at_max=rate(climate="High", time="Low"),
    ),

    # MOD 16a - Mandates vs. incentives - Next Gen Roadmapping
    ChoiceMetric(
        id="nextgen_policy_instrument",
        label="Policy instrument",
        options={
            "None": _BASE,
            "Incentive": rate(climate="Medium", financial="Low", time="High"),
            "Mandate": rate(climate="High", financial="Medium", time="Low"),
        },
        default_option="None",
    ),

    # MOD 16b - Decarbonisation commitment by stakeholders - Next Gen Roadmapping
    PercentageMetric(
        id="nextgen_decarb_commitment",
        label="Decarbonisation Commitment by Stakeholders (%)",
        impact_at_0=rate(climate="Very Low", financial="Very High"),
        impact_at_100=rate(climate="Very High", financial="Very Low"),
        default_pct=50.0,
    ),

    # MOD 17a - Priority on New Technology Notification - Next Gen Roadmapping
    PercentageMetric(
        id="nextgen_tech_priority",
        label="Priority on New Technology Notification (%)",
        impact_at_0=rate(climate="Very Low", financial="Very High", time="Very High"),
        impact_at_100=rate(climate="Very High", financial="Very Low", time="Very Low"),
        default_pct=50.0,
    ),

    # MOD 17b - Technology toggle - Next Gen Roadmapping
    ChoiceMetric(
        id="nextgen_tech_toggle",
        label="Priority on New Technology Notification",
        options={
            "SAF": rate(climate="Medium", financial="High", time="High"),
            "H2": rate(climate="Very High", financial="Medium", time="Low"),
            "Electric": rate(climate="Very High", financial="Medium", time="Low"),
            "Other": _BASE,
        },
        default_option="SAF",
    ),

    # MOD 18 - Actions needed by stakeholder - Next Gen Roadmapping
    ChoiceMetric(
        id="nextgen_actions",
        label="Actions Needed by Stakeholder",
        options={
            "Airlines": rate(climate="Low", financial="Low", time="High"),
            "OEMs": rate(climate="High", financial="Low", time="High"),
            "Government": rate(climate="High", financial="High", time="Low"),
            "Other": _BASE,
        },
        default_option="Airlines",
    ),

    # MOD 19 - Funding by Stakeholder - Extra
    # Shares must sum to 100%. Climate-neutral: only financial & time are rated,
    # so climate is omitted (not applicable) for every group. TODO: tune ratings.
    AllocationMetric(
        id="funding_split",
        label="Who pays for the transition",
        groups={
            "Government": rate(financial="Medium", time="High"),    # TODO
            "Private": rate(financial="High", time="Low"),          # TODO
            "NGO": rate(financial="Low", time="Medium"),            # TODO
            "Public": rate(financial="Low", time="Medium"),         # TODO
        },
        default_allocation={"Government": 25, "Private": 25, "NGO": 25, "Public": 25},
    ),
]


def build_simulation():
    """Return a ready-to-use :class:`~back_end.engine.Simulation` for the board.

    Weights are equal thirds by default — retune in the ScoreWeights below.
    """
    from .engine import Simulation  # local import to avoid a cycle

    return Simulation(
        metrics=list(MODULES),
        weights=ScoreWeights(climate=1.0, financial=1.0, time=1.0),  # TODO: retune
    )

File: front_end/js/modules.js
/* ============================================================================
 * modules.js — scenario configuration, mirroring back_end/modules.py
 *
 * ┌─────────────────────────────────────────────────────────────────────────┐
 * │  ✏️  THE ONLY FILE YOU NEED TO EDIT TO TUNE THE GAME.                    │
 * │                                                                         │
 * │  Ratings use rate(climate, financial, time) — POSITIONAL, unlike the    │
 * │  Python file which allows keywords. Same values, different syntax:      │
 * │                                                                         │
 * │      Python:  impact_at_0=rate(climate="Low", financial="Very High")    │
 * │      JS:      impact_at_0: rate("Low", "Very High"),                    │
 * │                                                                         │
 * │  Order is always climate, financial, time. Leave a trailing dimension   │
 * │  out (or pass null) to mark it "not applicable".                        │
 * │  Scale: "Very Low" 0 · "Low" 25 · "Medium" 50 · "High" 75 ·             │
 * │         "Very High" 100 — or any raw 0-100 number.                      │
 * │  Note the ':' after each property name — '=' breaks JavaScript.         │
 * │  Keep this file and back_end/modules.py in sync.                        │
 * └─────────────────────────────────────────────────────────────────────────┘
 * ========================================================================== */

"use strict";

// Baseline used by the "None"/"Other" options.
const _BASE = rate("Very Low", "Very Low", "Very Low");   // a 'do nothing' low end

const MODULES = [

  // MOD 01 - CO2 cuts - SAF
  new PercentageMetric({
    id: "saf_co2_cuts",
    label: "SAF CO2 Cuts (%)",
    impact_at_0: rate("Low", "Very High"),    // time not applicable
    impact_at_100: rate("Very High", "Medium"),  // time not applicable
    default_pct: 50.0,
  }),

  // MOD 02 - Time to 100% SAF - SAF
  new PercentageMetric({
    id: "saf_time_to_100",
    label: "Time to 100% SAF (% vol)",
    impact_at_0: rate("Very Low", "Medium", "Very Low"),
    impact_at_100: rate("Very High", "Medium", "Very High"),
    default_pct: 50.0,
  }),

  // MOD 03 - Willingness to contribute financially - SAF
  new PercentageMetric({
    id: "saf_WTC",
    label: "Willingness to contribute financially (% premium accepted)",
    impact_at_0: rate("Low", "Low"),          // time not applicable
    impact_at_100: rate("High", "High"),      // time not applicable
    default_pct: 50.0,
  }),

  // MOD 04 - System maintenance effort - Bio-Sensors
  // Level 1 = Low effort … Level 5 = High effort. Rate what each end MEANS
  // for climate / financial / time in the ratings below.
  new LevelMetric({
    id: "biosensor_maintenance",
    label: "System Maintenance Effort",
    impact_at_min: rate("Very Low", "Low"),   // time not applicable
    impact_at_max: rate("Very High", "High"), // time not applicable
  }),

  // MOD 05 - Reliability gain vs. current sensors - Bio-Sensors
  new PercentageMetric({
    id: "biosensor_reliability",
    label: "Reliability Gain vs. Current Sensors (% more reliable)",
    impact_at_0: rate("Very Low", "Very Low", "High"),
    impact_at_100: rate("Very High", "Very High", "Low"),
    default_pct: 50.0,
  }),

  // MOD 06 - Certification support - Bio-Sensors
  new LevelMetric({
    id: "biosensor_certification",
    label: "Certification Support",
    impact_at_min: rate("Very Low", null, "Very Low"),   // financial not applicable
    impact_at_max: rate("Very High", null, "Very High"), // financial not applicable
  }),

  // MOD 07 - Data Depth - Digital Passport
  new PercentageMetric({
    id: "passport_digital_depth",
    label: "Digital Depth (%)",
    impact_at_0: rate("Low", "High", "Low"),
    impact_at_100: rate("High", "Medium", "High"),
    default_pct: 50.0,
  }),

  // MOD 08 - Regulatory Support - Digital Passport
  new LevelMetric({
    id: "passport_reg_support",
    label: "Regulatory Support",
    impact_at_min: rate("Medium", "Very Low", "Very Low"),
    impact_at_max: rate("High", "Very High", "Very High"),
  }),

  // MOD 09 - Industry Adoption - Materials Digital Passport
  new LevelMetric({
    id: "passport_adoption",
    label: "Industry Adoption",
    impact_at_min: rate("Low", "Low", "Very Low"),
    impact_at_max: rate("High", "High", "Very High"),
  }),

  // MOD 10 - Funding payback period — REMOVED: replaced by MOD 19 (funding_split)
  // in the Emission Guide & Eco-Label group, per design decision.

  // MOD 11 - Global Participation (incl. eco-labels) - Emission Guide & Eco-label
  new LevelMetric({
    id: "EGE_participation",
    label: "Global Participation (incl. Eco-labels)",
    impact_at_min: rate("Very Low", "Very Low", "Very High"),
    impact_at_max: rate("Very High", "Very High", "Very Low"),
  }),

  // MOD 12 - Market trust in label - Emission Guide & Eco-label
  new LevelMetric({
    id: "EGE_market_trust",
    label: "Market Trust in Label",
    impact_at_min: rate("Low", null, "Very Low"),   // financial not applicable
    impact_at_max: rate("High", null, "Very High"), // financial not applicable
  }),

  // MOD 13 - Sustainability mode shift - Transport of Aircraft Elements
  new PercentageMetric({
    id: "transport_mode_shift",
    label: "Shift to Low Climate, Resource, and Air-Quality Impact Modes (%)",
    impact_at_0: rate("Very Low", "Very Low", "Very Low"),
    impact_at_100: rate("Very High", "Very High", "Very High"),
    default_pct: 50.0,
  }),

  // MOD 14 - Carbon tax - Transport of Aircraft Elements
  new PercentageMetric({
    id: "transport_carbon_tax",
    label: "Carbon Tax (%)",
    impact_at_0: rate("Low", "High", "Very High"),
    impact_at_100: rate("High", "Low", "Very Low"),
    default_pct: 50.0,
  }),

  // MOD 15 - Schedule buffer - Transport of Aircraft Elements
  new LevelMetric({
    id: "transport_buffer",
    label: "Schedule Buffer",
    impact_at_min: rate("Very Low", null, "High"),  // financial not applicable
    impact_at_max: rate("High", null, "Low"),       // financial not applicable
  }),

  // MOD 16a - Mandates vs. incentives - Next Gen Roadmapping
  new ChoiceMetric({
    id: "nextgen_policy_instrument",
    label: "Policy instrument",
    options: {
      "None": _BASE,
      "Incentive": rate("Medium", "Low", "High"),
      "Mandate": rate("High", "Medium", "Low"),
    },
    default_option: "None",
  }),

  // MOD 16b - Decarbonisation commitment by stakeholders - Next Gen Roadmapping
  new PercentageMetric({
    id: "nextgen_decarb_commitment",
    label: "Decarbonisation Commitment by Stakeholders (%)",
    impact_at_0: rate("Very Low", "High"),      // time not applicable
    impact_at_100: rate("Very High", "Medium"), // time not applicable
    default_pct: 50.0,
  }),

  // MOD 17a - Priority on New Technology Notification - Next Gen Roadmapping
  new PercentageMetric({
    id: "nextgen_tech_priority",
    label: "Priority on New Technology Notification (%)",
    impact_at_0: rate("Very Low", "Very High", "Very High"),
    impact_at_100: rate("Very High", "Very Low", "Very Low"),
    default_pct: 50.0,
  }),

  // MOD 17b - Technology toggle - Next Gen Roadmapping
  new ChoiceMetric({
    id: "nextgen_tech_toggle",
    label: "Priority on New Technology Notification",
    options: {
      "SAF": rate("Medium", "High", "High"),
      "H2": rate("Very High", "Medium", "Low"),
      "Electric": rate("Very High", "Medium", "Low"),
      "Other": _BASE,
    },
    default_option: "SAF",
  }),

  // MOD 18 - Actions needed by stakeholder - Next Gen Roadmapping
  new ChoiceMetric({
    id: "nextgen_actions",
    label: "Actions Needed by Stakeholder",
    options: {
      "Airlines": rate("Low", "Low", "High"),
      "OEMs": rate("High", "Low", "High"),
      "Government": rate("High", "High", "Low"),
      "Other": _BASE,
    },
    default_option: "Airlines",
  }),

  // MOD 19 - Funding by Stakeholder - Extra
  // Shares must sum to 100%. Climate-neutral: only financial & time are rated,
  // so climate is omitted (null / not applicable) for every group.
  new AllocationMetric({
    id: "funding_split",
    label: "Who pays for the transition",
    groups: {
      "Government": rate(null, "Medium", "High"),
      "Private":    rate(null, "High",   "Low"),
      "NGO":        rate(null, "Low",    "Medium"),
      "Public":     rate(null, "Low",    "Medium"),
    },
    default_allocation: { "Government": 25, "Private": 25, "NGO": 25, "Public": 25 },
  }),
];

// Weights are equal thirds by default — retune here.  // TODO: retune
function buildSimulation() {
  return new Simulation(MODULES, new ScoreWeights(1.0, 1.0, 1.0));
}
